[docs] TestCafe Roadmap 2021 (#5960)
DevExpress/testcafe: docs/articles/roadmap/README.md

Feature                                                                                           | Status
------------------------------------------------------------------------------------------------- | -------------------

> [HTTP API Testing](https://github.com/DevExpress/testcafe/issues/4288)                            | Planned

Reviewer: Should we merge https://github.com/DevExpress/testcafe/issues/2397 with https://github.com/DevExpress/testcafe/issues/4288? I think we don't need them both

Reviewer 2: I closed #2397.

[HTTP2/HTTP3 Support](https://github.com/DevExpress/testcafe-hammerhead/issues/1960)              | Planned

> [Set HTTP Cookies](https://github.com/DevExpress/testcafe/issues/4063)                            | Planned

Reviewer: What about `Interactive demo center`? Don't we want to include it? @kirovboris

Reviewer 3: Yes, we were going to include.

Author: [The new issue](https://github.com/DevExpress/testcafe/issues/5966)

## API

Feature                                                                                           | Status
------------------------------------------------------------------------------------------------- | -------------------
[Testing in Multiple Browser Windows](https://github.com/DevExpress/testcafe/issues/912)          | **Implemented**

> [Dedicated *Scroll* Action](https://github.com/DevExpress/testcafe/issues/527)                    | Planned

Reviewer: We have at least 3 feature requests for the scroll action https://github.com/DevExpress/testcafe/issues/4399 https://github.com/DevExpress/testcafe/issues/1681 https://github.com/DevExpress/testcafe/issues/1256
But all of them are closed so let's create a new one.

> [Dedicated *Scroll* Action](https://github.com/DevExpress/testcafe/issues/527)                    | Planned

Author: [The new issue](https://github.com/DevExpress/testcafe/issues/5965)

> [View Active Cookies](https://github.com/DevExpress/testcafe/issues/4428)                         | Planned

Reviewer 2: It's not `active` cookie.

I propose the the `Inspect cookies` term.

> [View Active Cookies](https://github.com/DevExpress/testcafe/issues/4428)                         | Planned

Reviewer: Ok, let's call it `Cookie management` if you don't mind

> [View Active Cookies](https://github.com/DevExpress/testcafe/issues/4428)                         | Planned

Author: ok, `Cooke Management` it is. Since there is no issue for both get and set, I've made a new one: [Cookie Management](https://github.com/DevExpress/testcafe/issues/5964)

## Debugging

Feature                                                                                           | Status
------------------------------------------------------------------------------------------------- | -------------------
[Selector Debug Panel](https://github.com/DevExpress/testcafe/issues/3244)                        | **In Progress**

## Test Stability

Feature                                                                                           | Status
------------------------------------------------------------------------------------------------- | -------------------
[Cache Page Assets Between Tests](https://github.com/DevExpress/testcafe-hammerhead/issues/863)   | **In Progress**

> Improve Waiting Mechanism                                                                         | Planned

Reviewer: I'm not sure about this statement. Can we give something more specific? @miherlosev

Reviewer 2: `Cache Page Assets Between Tests` is strict and technical term. For roadmap, I propose to use the more common term - `HTTP caching`

Reviewer: I'm sorry, my comment was for `Improve Waiting Mechanism`

Reviewer: @AndreyBelym what do you think? you mentioned timeouts AFAIR

Author: I've added it in and created a [new issue](https://github.com/DevExpress/testcafe/issues/5967)

> Enhance TestCafe HTTPS Capabilities                                                               | Planned

Reviewer 2: Please create an issue for this task.

Reviewer: 1. Implement using the default https protocol
2. Simplify ssl certificate setup

Author: split in two and created issues

Feature                                                                                           | Status
------------------------------------------------------------------------------------------------- | -------------------

> [Testing in Multiple Browser Windows](https://github.com/DevExpress/testcafe/issues/912)          | **Implemented**

Reviewer 4: ```suggestion
[Multiple Window Support](https://github.com/DevExpress/testcafe/issues/912)          | **Implemented**
```

[Selector Debug Panel](https://github.com/DevExpress/testcafe/issues/3244)                        | **In Progress**
[HTTP API Testing](https://github.com/DevExpress/testcafe/issues/4288)                            | Planned
[Cookie Management](https://github.com/DevExpress/testcafe/issues/5964)                           | Planned

## API

Feature                                                                                           | Status
------------------------------------------------------------------------------------------------- | -------------------
[Dedicated *Scroll* Action](https://github.com/DevExpress/testcafe/issues/5965)                   | Planned

## Debugging

Feature                                                                                           | Status
------------------------------------------------------------------------------------------------- | -------------------

> [Interactive Demo Center](https://github.com/DevExpress/testcafe/issues/5966)                     | Planned

Reviewer 4: ```suggestion
[Interactive Demo Page](https://github.com/DevExpress/testcafe/issues/5966)                     | Planned
```

Author: We already have an [interactive demo page](https://devexpress.github.io/testcafe/example/), don't we?

Reviewer 4: It's an example page, not a demo page. It doesn't demonstrate anything on its own.

## Test Stability

Feature                                                                                           | Status
------------------------------------------------------------------------------------------------- | -------------------

> [Retry Failed Requests for Tested Webpages](https://github.com/DevExpress/testcafe/pull/5738)     | **Implemented**

Reviewer 4: ```suggestion
[Automatically Retry Failed Requests](https://github.com/DevExpress/testcafe/pull/5738)     | **Implemented**
```

Author: Not all requests, just an initial request for the page (when we request HTML from a server)

Reviewer 4: Your phrasing doesn't make this clear.

```suggestion
[Automatically Reissue Failed Page Requests](https://github.com/DevExpress/testcafe/pull/5738)     | **Implemented**
```

or 

```suggestion
[Repeat Page Requests on Failure](https://github.com/DevExpress/testcafe/pull/5738)     | **Implemented**
```

[HTTP Caching](https://github.com/DevExpress/testcafe-hammerhead/issues/863)                      | **In Progress**

> [Improve Waiting Mechanism for Assertions](https://github.com/DevExpress/testcafe/issues/5967)    | Planned

Reviewer 4: please please please let's stop using the word "mechanism", native English speakers don't use it the way Russian speakers do.

```suggestion
[Improve Assertion Waiting Algorithm](https://github.com/DevExpress/testcafe/issues/5967)    | Planned
```

Author: I agree, `mechanism` isn't perfect, but we [use it a lot ](https://devexpress.github.io/testcafe/documentation/guides/concepts/built-in-wait-mechanisms.html)in our docs. I'd leave it as is for consistency (until we reconsider the term, which we should).
`Algorithm` doesn't fit the context in my opinion

Reviewer 4: a) could you please elaborate re: 'algorithm'? 
here's the Oxford dictionary [definition](https://www.oxfordlearnersdictionaries.com/definition/english/algorithm) of the word: 
`​a set of rules that must be followed when solving a particular problem`
what are we going to improve if not the steps that are being followed when determining the timeout value?

b) we can just leave it out if you still don't agree:

```suggestion
[Improve the Quality of Automatic Pre-Assertion Timeouts](https://github.com/DevExpress/testcafe/issues/5967)    | Planned
```

> [Enable HTTPS connection Between the Proxy and Browsers](https://github.com/DevExpress/testcafe/issues/5974)| Planned

Reviewer 4: either add the definite article or make it plural
```suggestion
[Enable HTTPS connections Between the Proxy and the Browsers](https://github.com/DevExpress/testcafe/issues/5974)| Planned
```

Author: `Browsers` doesn't need an article here (talking about all browsers out there)

```suggestion
[Enable HTTPS connections Between Browsers and the Proxy](https://github.com/DevExpress/testcafe/issues/5974)| Planned
```

> [Simplify SSL Certificate Setup](https://github.com/DevExpress/testcafe/issues/5973)              | Planned

Reviewer 4: ```suggestion
[Simplify the SSL Workflow](https://github.com/DevExpress/testcafe/issues/5973)              | Planned
```
The word 'setup' might be a bit misleading - TestCafe can not help users set up SSL certificates for their websites

> [Skip JS Errors on Chosen Pages](https://github.com/DevExpress/testcafe/issues/2775)              | Planned

Reviewer 4: Pages chosen by whom?

```suggestion
[Ignore JS Errors On a Per-Page Basis](https://github.com/DevExpress/testcafe/issues/2775)              | Planned
```
